### ISSUE TITLE
Performance issues in DBTransactionStorage

### ISSUE THREAD

**Issue by remmeier** [corda/corda#5831]
The DBTransactionStorage class is central when using recordTransactions to save transactions in the vault. Within a unit test and Corda 4.1 we noticed it takes about 500ms to store 1000 transactions. We managed to reduce that number to 150ms. After an upgrade to most recent Corda 4.3 resp. Corda 4.4 master we noticed it to take 1500ms.

There seem to be three bottlenecks:

- Corda 4.3 introduced `updateTransaction` backported from Corda Enterprise which is a cause of additional work. Most importantly, the `updateTransaction` is executed first, whereas a further check for the existence is executed second, causing redudant work.
- `NodeVaultService` makes use of `val netUpdate = updates.reduce { update1, update2 -> update1 + update2 }` to aggregate all updates. With n transactions to save, in every step two transactions will be merged resulting in new data structures holding all its touched states. This in turn leads to a quadratic runtime. When attempting to store a few hundret or thousand transactions at a time, it will consume most of the CPU time.
- while the vault service takes as argument a list of transaction, the underlying DBTransactionStorage only takes individual transactions with `override fun addTransaction(transaction: SignedTransaction)`. There are many sequential reads and writes to the database, which can be replaced by much faster batch operations. 

will have a fix ready shortly if there is an interest. we do batching to in-source existing non-ledger data into Corda. thereby performance is quite essential due to the large amount of data.

**Comment by davidrapacchiale** [comment]
Thanks @remmeier . I have created three different bugs in our OS Jira project to track this issue.
1. https://r3-cev.atlassian.net/browse/CORDA-3545
2. https://r3-cev.atlassian.net/browse/CORDA-3546
3. https://r3-cev.atlassian.net/browse/CORDA-3547

It would be good to understand what's your proposed fix for this. Does it tackle all the above?

**Comment by remmeier** [comment]
yes, https://github.com/corda/corda/pull/5860 addresses the three above. But for details see the comments in the PR/commits. I tried to minimize changes to the existing source code.

Last piece of the puzzle would be https://github.com/corda/corda/issues/5725. I solved 50% with https://github.com/corda/corda/pull/5841. Still open is avoiding the redudant storage of the schema. Meaning that a transaction should not be stored as single blob in the DB, but normalized by actual data, schema, and potentially signatures. That will both speed-up DBTransactionStorage in terms of CPU load (serialization and compression is expensive on large blobs) and reduce DB storage requirement rather drastically.  That part I intend to address the coming days if there is an interest. 

We are quite close in getting our system running with an expected 50 mio transactions in the early days. It is in progress that I can share a little example app and more things the coming weeks.
